### ISSUE TITLE
`List` incorrect swap behavior of front and back

### ISSUE THREAD

**Issue by Xrayez** [godotengine/godot#41981]
**Godot version:**
Godot 3.2, master @ e49d25bac08d1205f5c2cd7126a0323fd028e0cc.

**OS/device including version:**
Windows 10.

**Issue description:**
Swapping `List` front and back elements does not seem to swap the contents of those elements (C++ only):

```cpp
List<Variant> list;
Variant str = "Godot";
List<Variant>::Element *n_str = list.push_back(str);
Variant color = Color(0, 0, 1);
List<Variant>::Element *n_color = list.push_back(color);

CHECK(list.front()->get() == "Godot"); // OK.
CHECK(list.back()->get() == Color(0, 0, 1)); // OK.

list.swap(n_str, n_color);

CHECK(list.front()->get() == Color(0, 0, 1)); // Fails.
CHECK(list.back()->get() == "Godot"); // Fails.
```

Note that this seems to work with swapping elements in the middle of the list.

**Steps to reproduce:**
1. Compile the engine with tests: `scons tests=yes`.
2. Run tests with `godot --test`.

**Minimal reproduction project:**

**Comment by akien-mga** [comment]
Indeed, `List::swap` changes `next_ptr` and `prev_ptr` for both elements, but doesn't seem to impact `data->first` and `data->last`:

https://github.com/godotengine/godot/blob/e49d25bac08d1205f5c2cd7126a0323fd028e0cc/core/list.h#L393-L420

https://github.com/godotengine/godot/blob/e49d25bac08d1205f5c2cd7126a0323fd028e0cc/core/list.h#L174-L200
